go/tasks/plugins/array/k8s/monitor.go

```diff
@@ -24,73 +23,100 @@ import (
 	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/flytek8s"
 
 	idlCore "github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
-	"github.com/lyft/flytestdlib/errors"
+	errors2 "github.com/lyft/flytestdlib/errors"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
 
 	"github.com/lyft/flyteplugins/go/tasks/logs"
 	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
 )
 
 const (
-	ErrCheckPodStatus errors.ErrorCode = "CHECK_POD_FAILED"
+	ErrCheckPodStatus errors2.ErrorCode = "CHECK_POD_FAILED"
 )
 
-func CheckSubTasksState(ctx context.Context, tCtx core.TaskExecutionContext, kubeClient core.KubeClient,
-	dataStore *storage.DataStore, outputPrefix, baseOutputDataSandbox storage.DataReference, currentState *arrayCore.State) (
+func LaunchAndCheckSubTasksState(ctx context.Context, tCtx core.TaskExecutionContext, kubeClient core.KubeClient,
+	config *Config, dataStore *storage.DataStore, outputPrefix, baseOutputDataSandbox storage.DataReference, currentState *arrayCore.State) (
 	newState *arrayCore.State, logLinks []*idlCore.TaskLog, err error) {
+	if int64(currentState.GetExecutionArraySize()) > config.MaxArrayJobSize {
+		ee := fmt.Errorf("array size > max allowed. Requested [%v]. Allowed [%v]", currentState.GetExecutionArraySize(), config.MaxArrayJobSize)
+		logger.Info(ctx, ee)
+		currentState = currentState.SetPhase(arrayCore.PhasePermanentFailure, 0).SetReason(ee.Error())
+		return currentState, logLinks, nil
+	}
 
 	logLinks = make([]*idlCore.TaskLog, 0, 4)
 	newState = currentState
-
 	msg := errorcollector.NewErrorMessageCollector()
-	newArrayStatus := arraystatus.ArrayStatus{
+	newArrayStatus := &arraystatus.ArrayStatus{
 		Summary:  arraystatus.ArraySummary{},
 		Detailed: arrayCore.NewPhasesCompactArray(uint(currentState.GetExecutionArraySize())),
 	}
 
+	// If we have arrived at this state for the first time then currentState has not been
+	// initialized with number of sub tasks.
+	if len(currentState.GetArrayStatus().Detailed.GetItems()) == 0 {
+		currentState.ArrayStatus = *newArrayStatus
+	}
+
 	for childIdx, existingPhaseIdx := range currentState.GetArrayStatus().Detailed.GetItems() {
 		existingPhase := core.Phases[existingPhaseIdx]
+		indexStr := strconv.Itoa(childIdx)
+		podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
 		if existingPhase.IsTerminal() {
 			// If we get here it means we have already "processed" this terminal phase since we will only persist
 			// the phase after all processing is done (e.g. check outputs/errors file, record events... etc.).
+
+			// Since we know we have already "processed" this terminal phase we can safely deallocate resource
+			err = deallocateResource(ctx, tCtx, config, childIdx)
```

Comment:
I think this can have an issue, in the last round (where SummaryToPhase would return a different phase), we will not dealocate resources for the tasks that JUST completed.

Reply:
I see. But we can run into issues if we call after the task just completed as well. Any ideas on how to handle this?

Reply:
Actually shouldn't finalized be call before returning so it would deallocate the resource there in this scenario?

Reply:
ic. I missed that you attempt to deallocate tokens in finalize anyway... awesome